### PR TITLE
Add build_id to module identification logic

### PR DESCRIPTION
**Pull request by dimitry-**
Modules are now identified by path and build_id instead of
being identified by path alone.

Bug: http://b/183381209
Bug: http://b/183380723
Test: Start orbit, instrument functions, caprue, inspect callstacks,
      callgraphs, frame tracks. Select/unselect functions.

### CODE REVIEW

**Comment by antonrohr** [comment]
I am not done with reviewing yet, but I have an early request. Since you are touching `ModuleManager::AddOrUpdateModules`, the function calls `ModuleData::UpdateIfChanged`. Can you also update this function (Maybe its not needed anymore at all).

**Comment by dimitry-** [comment]
> I am not done with reviewing yet, but I have an early request. Since you are touching `ModuleManager::AddOrUpdateModules`, the function calls `ModuleData::UpdateIfChanged`. Can you also update this function (Maybe its not needed anymore at all).

done.

File: src/OrbitClientModel/CaptureSerializer.cpp [hunk 1]
```diff
@@ -92,19 +83,35 @@ CaptureInfo GenerateCaptureInfo(
   process->set_command_line(capture_data.process()->command_line());
   process->set_is_64_bit(capture_data.process()->is_64_bit());
 
+  absl::flat_hash_map<std::string, std::string> module_path_to_build_id;
   for (const auto& [module_path, memory_space] : capture_data.process()->GetMemoryMap()) {
```
Comment (antonrohr) [inline src/OrbitClientModel/CaptureSerializer.cpp:87]:
```suggestion
  for (const auto& [module_path, module_in_memory] : capture_data.process()->GetMemoryMap()) {
```

File: src/OrbitClientModel/CaptureSerializer.cpp [hunk 2]
```diff
@@ -92,19 +83,35 @@ CaptureInfo GenerateCaptureInfo(
...
     module_info->set_file_size(module->file_size());
-    module_info->set_address_start(memory_space.start);
-    module_info->set_address_end(memory_space.end);
+    module_info->set_address_start(memory_space.start());
+    module_info->set_address_end(memory_space.end());
     module_info->set_build_id(module->build_id());
     module_info->set_load_bias(module->load_bias());
   }
 
+  for (const auto& pair : capture_data.instrumented_functions()) {
```
Comment (antonrohr) [inline src/OrbitClientModel/CaptureSerializer.cpp:102]: pers pref: 
```suggestion
  for (const auto& [absolute_address, instrumented_function] : capture_data.instrumented_functions()) {
```

File: src/OrbitClientModel/CaptureSerializer.cpp [hunk 3]
```diff
@@ -92,19 +83,35 @@ CaptureInfo GenerateCaptureInfo(
...
+    const std::string& module_path = pair.second.file_path();
+    CHECK(module_path_to_build_id.contains(module_path));
```
Comment (antonrohr) [inline src/OrbitClientModel/CaptureSerializer.cpp:104]: I might miss something, but shouldn't an `InstrumentedFunction` (`pair.second`) now contain a file_build_id? This would make the `module_path_to_build_id` map unnecessary

File: src/OrbitGl/LiveFunctionsDataView.cpp [hunk 4]
```diff
@@ -406,8 +406,9 @@ void LiveFunctionsDataView::OnDataChanged() {
   const absl::flat_hash_map<uint64_t, orbit_grpc_protos::InstrumentedFunction>&
       instrumented_functions = app_->GetCaptureData().instrumented_functions();
   for (const auto& pair : instrumented_functions) {
```
Comment (antonrohr) [inline src/OrbitGl/LiveFunctionsDataView.cpp:408]: pers pref: while you are at it
```suggestion
  for (const auto& [unused, instrumented_function] : instrumented_functions) {
```

File: src/OrbitGl/ModulesDataView.cpp [hunk 5]
```diff
@@ -195,7 +195,8 @@ void ModulesDataView::UpdateModules(const ProcessData* process) {
   modules_.clear();
   module_memory_.clear();
   for (const auto& [module_path, memory_space] : process->GetMemoryMap()) {
```
Comment (antonrohr) [inline src/OrbitGl/ModulesDataView.cpp:197]:
```suggestion
  for (const auto& [module_path, module_in_memory] : process->GetMemoryMap()) {
```

**Review by antonrohr — COMMENTED** [review]
I like the change a lot! I have some comments, but in general it looks very good

In ModuleManagerTest, please update the names of the tests (`GetModuleByPath` -> `GetModuleByPathAndBuildId` and `GetMutableModuleByPath` -> `GetMutableModuleByPathAndBuildId`) 

Also I think an additional test with an empty build_id for `ModuleManager::AddOrUpdateModule` is not a bad idea.

File: src/OrbitClientData/include/OrbitClientData/ProcessData.h [hunk 6]
```diff
@@ -59,25 +67,35 @@ class ProcessData final {
...
   }
-  const absl::flat_hash_map<std::string, MemorySpace>& GetMemoryMap() const {
+  const absl::node_hash_map<std::string, ModuleInMemory>& GetMemoryMap() const {
     return module_memory_map_;
   }
-  bool IsModuleLoaded(const std::string& module_path) const {
-    return module_memory_map_.contains(module_path);
+
+  [[nodiscard]] const ModuleInMemory* FindModuleByPath(const std::string& module_path) const {
+    auto it = module_memory_map_.find(module_path);
```
[the remark this answers is not in the document]
Comment (antonrohr) [reply src/OrbitClientData/include/OrbitClientData/ProcessData.h:82]: I meant the body of the method

File: src/OrbitClientData/ModuleData.cpp [hunk 7]
```diff
@@ -26,20 +26,21 @@ void ModuleData::UpdateIfChanged(ModuleInfo info) {
   absl::MutexLock lock(&mutex_);
 
   CHECK(file_path() == info.file_path());
+  CHECK(build_id() == info.build_id());
+
+  // This function only makes sense if build_id is empty
+  CHECK(build_id().empty());
 
   // TODO(171878807): Remove this as soon as a better way of distinguishing modules is implemented.
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientData/ModuleData.cpp]: done.

File: src/OrbitClientData/ModuleDataTest.cpp [hunk 8]
```diff
@@ -229,11 +225,49 @@ TEST(ModuleData, UpdateIfChanged) {
...
+  EXPECT_TRUE(module.GetFunctions().empty());
+
+  // We cannot change a module with non-empty build_id
+  module_info.set_name("different name");
+  EXPECT_DEATH(module.UpdateIfChanged(module_info), "Check failed");
+
+  // adding symbols should work.
+  ModuleSymbols symbols;
+  module.AddSymbols(symbols);
+  EXPECT_TRUE(module.is_loaded());
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientData/ModuleDataTest.cpp:272]: it is, EXPECT_DEATH forks, the main process continues execution after checking that the child died.

File: src/OrbitClientData/include/OrbitClientData/ProcessData.h [hunk 9]
```diff
@@ -18,20 +20,26 @@
...
 #include "module.pb.h"
 #include "process.pb.h"
 #include "symbol.pb.h"
 
 // Small struct to model a space in memory occupied by a module.
-struct MemorySpace {
-  explicit MemorySpace(uint64_t start, uint64_t end) : start(start), end(end) {}
-  uint64_t start;
-  uint64_t end;
+class ModuleInMemory {
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientData/include/OrbitClientData/ProcessData.h:28]: added path_, changed FindModuleByAddress accordingly; I think I like ModuleInMemory a bit better.

File: src/OrbitClientData/include/OrbitClientData/ProcessData.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientData/include/OrbitClientData/ProcessData.h]: done.

File: src/OrbitClientModel/CaptureData.cpp [hunk 10]
```diff
@@ -159,6 +161,14 @@ const FunctionInfo* CaptureData::FindFunctionByModulePathAndOffset(const std::st
...
 }
 
+const std::string& CaptureData::FindModuleBuildIdByAddress(uint64_t absolute_address) const {
+  const ModuleData* module_data = FindModuleByAddress(absolute_address);
+  if (module_data == nullptr) {
+    return kEmptyBuildId;
+  }
+  return module_data->build_id();
+}
+
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientModel/CaptureData.cpp]: makes sense. done.

File: src/OrbitClientModel/CaptureSerializer.cpp (hunk 1 above)
Comment (dimitry-) [reply src/OrbitClientModel/CaptureSerializer.cpp]: done.

File: src/OrbitClientModel/CaptureSerializer.cpp (hunk 3 above)
Comment (dimitry-) [reply src/OrbitClientModel/CaptureSerializer.cpp]: right. thanks for catching this - fixed.

File: src/OrbitClientModel/CaptureSerializer.cpp (hunk 2 above)
Comment (dimitry-) [reply src/OrbitClientModel/CaptureSerializer.cpp]: done.

File: src/OrbitGl/LiveFunctionsDataView.cpp (hunk 4 above)
Comment (dimitry-) [reply src/OrbitGl/LiveFunctionsDataView.cpp]: done

File: src/OrbitGl/ModulesDataView.cpp (hunk 5 above)
Comment (dimitry-) [reply src/OrbitGl/ModulesDataView.cpp]: done.

**Comment by dimitry-** [comment]
> In ModuleManagerTest, please update the names of the tests (`GetModuleByPath` -> `GetModuleByPathAndBuildId` and `GetMutableModuleByPath` -> `GetMutableModuleByPathAndBuildId`)

renamed

> Also I think an additional test with an empty build_id for `ModuleManager::AddOrUpdateModule` is not a bad idea.

I am not sure what you mean, AddOrUpdateModules is currently testing the function with and without build_id. Which test-case is missing there?

File: src/ClientProtos/capture_data.proto [hunk 11]
```diff
@@ -38,6 +38,7 @@ message FunctionInfo {
   string name = 1;
   string pretty_name = 2;
   string loaded_module_path = 3;
+  string loaded_module_build_id = 12;
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/ClientProtos/capture_data.proto:41]: added rename commit.

File: src/OrbitClientData/ModuleDataTest.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (antonrohr) [reply src/OrbitClientData/ModuleDataTest.cpp:275]: Good to know! Thanks

**Review by antonrohr — APPROVED** [review]
>I am not sure what you mean, AddOrUpdateModules is currently testing the function with and without build_id. Which test-case is missing there?

Sorry, my mistake. I did not read the test careful enough. 

Thank you for this change!

File: src/OrbitClientGgp/ClientGgp.cpp [hunk 12]
```diff
@@ -182,7 +182,12 @@ ErrorMessageOr<void> ClientGgp::LoadModuleAndSymbols() {
   OUTCOME_TRY(module_infos, process_client_->LoadModuleList(target_process_.pid()));
 
   LOG("List of modules");
+  std::string target_process_build_id;
   for (const ModuleInfo& info : module_infos) {
+    // TODO(dimitry): eventually we want to get build_id directly from ProcessData
```
Comment (florian-kuebler) [inline src/OrbitClientGgp/ClientGgp.cpp:187]: Should this be a bug?

File: src/OrbitClientModel/CaptureDeserializer.cpp [hunk 13]
```diff
@@ -160,10 +160,22 @@ ErrorMessageOr<CaptureListener::CaptureOutcome> LoadCaptureInfo(
...
-    const ModuleData* module_data =
-        module_manager->GetModuleByPath(function.second.loaded_module_path());
+    instrumented_function.set_file_path(function.second.module_path());
+    const ModuleData* module_data = module_manager->GetModuleByPathAndBuildId(
+        function.second.module_path(), function.second.module_build_id());
+
+    // In the case module_data was not found by build id check if FunctionInfo build_id is empty,
+    // in which case assume this is capture saved with Orbit v1.61 and below try to get module
+    // build id from path and do another lookup.
+    if (module_data == nullptr && function.second.module_build_id().empty()) {
```
Comment (florian-kuebler) [inline src/OrbitClientModel/CaptureDeserializer.cpp:170]: We should probably add a test for both cases?

**Review by florian-kuebler — APPROVED** [review]
LGTM. I would appreciate one or two additional tests, but that would be just a bonus. Thanks for the change!

File: src/OrbitClientGgp/ClientGgp.cpp (hunk 12 above)
Comment (dimitry-) [reply src/OrbitClientGgp/ClientGgp.cpp:187]: it is part of the draft CL (start_capture sends build_id for the executable).

**Review by dimitry- — COMMENTED** [review]
I will add tests in a follow up CL

File: src/OrbitClientData/ModuleManagerTest.cpp [hunk 14]
```diff
@@ -222,8 +226,8 @@ TEST(ModuleManager, GetOrbitFunctionsOfProcess) {
     ASSERT_EQ(functions.size(), 1);
 
     FunctionInfo& function{functions[0]};
-    EXPECT_EQ(function.name(), kOrbitNameMangled);
-    EXPECT_EQ(function.pretty_name(), kOrbitName);
+    EXPECT_EQ(function.name(), orbit_name_mangled);
+    EXPECT_EQ(function.pretty_name(), orbit_name);
     EXPECT_EQ(function.address(), 500);
   }
 }
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply src/OrbitClientData/ModuleManagerTest.cpp:233]: done.

File: src/OrbitClientModel/CaptureDeserializer.cpp (hunk 13 above)
Comment (dimitry-) [reply src/OrbitClientModel/CaptureDeserializer.cpp:170]: done.
